contracts/impls/SimplePriceOracle.sol

```diff
@@ -4,47 +4,93 @@ import "../interfaces/PriceOracleInterface.sol";
 import "../impls/PriceOracleConfig.sol";
 import "../libs/Percentage.sol";
 import "../roles/PriceFeederRole.sol";
+import "../libs/Arrays.sol";
 
-contract SimplePriceOracle is PriceOracleConfig, PriceFeederRole, PriceOracleInterface {
-    struct PriceData {
+library PriceOracleStructs {
+    struct PriceRecord {
         uint price;
         uint timestamp;
     }
+}
 
-    mapping(address => uint) private prices;
-    mapping(address => PriceData) private priceSnapshots;
-
-    bool public constant isPriceOracle = true;
+contract PriceOracleDataSource is PriceFeederRole {
+    // key => feeder => price record
+    mapping(address => mapping(address => PriceOracleStructs.PriceRecord)) private priceRecords;
+    // key => hasUpdate
+    mapping(address => bool) public hasUpdate;
 
-    event PriceUpdated(address indexed addr, uint price);
+    // to store temp non-expired records for `findMedianPrice` method
+    uint[] private validPrices;
```

Comment:
is it possible to remove this and use memory array instead?

Reply:
It is possible, but as memory arrays are fix sized, to get a filtered result array for finding median, extra copy operations and number of filtered items variable would be needed. I guess the cost (complexity and gas) is worthy?